cpp/src/arrow/compute/row/grouper.h

```diff
@@ -39,10 +82,19 @@ class ARROW_EXPORT Grouper {
   static Result<std::unique_ptr<Grouper>> Make(const std::vector<TypeHolder>& key_types,
                                                ExecContext* ctx = default_exec_context());
 
-  /// Consume a batch of keys, producing the corresponding group ids as an integer array.
+  /// Consume a batch of keys, producing the corresponding group ids as an integer array,
+  /// over a slice defined by an offset and length, which defaults to the batch length.
+  /// Currently only uint32 indices will be produced, eventually the bit width will only
+  /// be as wide as necessary.
+  virtual Result<Datum> Consume(const ExecSpan& batch, int64_t consume_offset = 0,
```

Comment:
What are these changes for?

Reply:
This supports consuming part of a batch at a time. The boundaries for consuming are determined by [segmentation](https://github.com/apache/arrow/pull/34311#issuecomment-1442318634). When a segment boundary is reached, the segmented aggregation can output the aggregation state accumulated thus far.